### ISSUE TITLE
User management

### ISSUE THREAD

**Issue by moifort** [jhipster/generator-jhipster#1525]
Hello the community,

I would like to know if it's will be interesting to implement in Jhipster the user management functionality. The idea is to provide for administrator user an interface to manage all the users.

We can start to implement some mains functions like:
- Display all user with summary informations
- Allow the admin user to modify setting for a specific user:
  - Change first name
  - Change last name
  - Change email
  - Activate/Disable an user
  - Change default language
  - Change authorities 

I already make a prototype for my application:

![capture d ecran 2015-05-24 a 23 04 54](https://cloud.githubusercontent.com/assets/4958079/7789674/5886e426-0269-11e5-94de-d19bdff4e8d5.png)

![capture d ecran 2015-05-24 a 23 05 07](https://cloud.githubusercontent.com/assets/4958079/7789677/719aaad8-0269-11e5-86ce-23d8f6b474e5.png)
 
Thibaut.

**Comment by atomfrede** [comment]
I like it because I'll need it in the future, so if there is something build in that would be great. But I think there was already a similar issue some time ago. Not sure what happened to it.

**Comment by moifort** [comment]
@atomfrede I make research, but I don't find an issue talking about. If you remember the keyword, I will update this issue, and delete this one.

**Comment by andidev** [comment]
:thumbsup:

**Comment by gomesp** [comment]
:+1: that would be awesome

**Comment by hirro** [comment]
[image: :+1:] So nice!

2015-05-25 7:40 GMT+02:00 Pablo <notifications@github.com>:

**Comment by wuethrich44** [comment]
:+1:

**Comment by jdubois** [comment]
Yes that would be great.
We've always wanted to have this, only we lacked the time to implement it.

**Comment by moifort** [comment]
@jdubois I can make a PR, but before we can discuss how to implement it. I am new in Jhipster and I don't know if I implement as well like you want.

**Comment by deepu105** [comment]
:+1:

**Comment by wuethrich44** [comment]
I suggest to add the user management under the admin navigation instead as a separate menu entry.

**Comment by moifort** [comment]
This is how I implemented the functionality in my prototype

## Server Side

`domain/User.java`
- I extends with `AbstractAuditingEntity` to have the creation date of the user. The problem is only mongoDB and SQL DB type can manage it, but not for cassandra (maybe other alternative?)
- I use `private Set<Authority> authorities` but same not implemented with cassandra (it's `private Set<String> authorities`, I don't know how hibernate work with cassandra)
- I remove `@Pattern(regexp = "^[a-z0-9]*$")` annotation rule, because when you want to update anonymousUser it's not working

I decided to create a special repository and ressource: `repository/UserManagmentRepository.java` and `web/rest/UserManagmentResource.java`. The reason why I don't use the existing files `repository/UserRepository.java` and `web/rest/UserResource.java` it's to not alter the default behavior and to separate the user actions like  authentification, reset password etc. to the user management actions, but maybe it's a bad idea.

`web/rest/UserManagmentResource.java`
- findOne with eager on authorities
- findAll with eager on authorities

`repository/UserManagmentRepository.java`
- getAll
- getUser
- update use a specific DTO `web/rest/dto/UserManagmentDTO.java`and use `recipe/service/UserService.java`to update data 

I modified `recipe/service/UserService.java` to update the data from UserManagment functionnality and create a new DTO.

## Client side
I use the same construction like jhispter.
- I create a new menu because the functionality in the administration menu it's really oriented system functionalities and not business functionalities. But maybe, put in the administrator menu as suggested @wuethrich44 
- I remove the text of the edit/display buttons to have more space for user data in the table

If you have any comments?

Thibaut.

**Comment by piedrahitac** [comment]
This would be great!

**Comment by PeterEltgroth** [comment]
:+1:

**Comment by MatthieuLebigre** [comment]
:+1:

**Comment by Qinwroks** [comment]
![](https://assets-cdn.github.com/images/icons/emoji/unicode/1f44d.png)

**Comment by atomfrede** [comment]
@moifort I think I'll need a user management in the next week. So before I'll implement something that will be thrown away: Is there any progress on your PR? Can I help you with something?

**Comment by moifort** [comment]
@atomfrede I work on it. It will be a pleasure if you can help me and advise me on what I do! I will push my code on my branch before the end of the week.

**Comment by amaillere** [comment]
+1

**Comment by barais** [comment]
:+1:

**Comment by Qinwroks** [comment]
it will be very useful! :+1:

**Comment by moifort** [comment]
This is a first version of user management functionality. You can download the generator on my branch https://github.com/moifort/generator-jhipster/tree/user-managment. It will only work with default value (java 8, mysql, etc.). @atomfrede I wait your feedbacks! 

You will find it in the menu `Administration` as said @wuethrich44:
![capture d ecran 2015-06-12 a 04 46 31](https://cloud.githubusercontent.com/assets/4958079/8122701/205d7b24-10be-11e5-9414-81463e96ebd6.png)

**Comment by atomfrede** [comment]
Great! Will try it later in the evening.

**Comment by matdere** [comment]
Cool !  Great feature!
Mathieu

**Comment by moifort** [comment]
![usermanagement](https://cloud.githubusercontent.com/assets/4958079/8134698/905ac53a-1131-11e5-8eae-3799af31f6ed.gif)

**Comment by piedrahitac** [comment]
That look really sweet! Definitely looking forward to this!

**Comment by PeterEltgroth** [comment]
:+1: I agree, that does look nice!

**Comment by atomfrede** [comment]
@moifort Sorry I don't think a can have closer look this until monday, but it looks pretty awesome, great work :+1:

**Comment by moifort** [comment]
Thank you! @atomfrede no problem I will try to progress on the code.

**Comment by atomfrede** [comment]
@moifort You have a few typos in your ``index.js``, such that you can't generate a project:  e.g. ``_userManagementDTO.java`` but it should be ``_UserManagementDTO.java``

**Comment by moifort** [comment]
My mistake! I will correct it (Find and replace action :disappointed:).

Do you see other things?

 By the way I try to work on the cassandra compatibility, and I would like to know how you test it? how you simulate a Cassandra or Mongo DB?

**Comment by piedrahitac** [comment]
@moifort Both MongoDB and Cassandra have official repos in the Docker registry, maybe you could take some images for a spin...

**Comment by moifort** [comment]
@piedrahitac Thank you!

**Comment by phouverneyuff** [comment]
@moifort , can I help you with something?

I'll clone your fork and take a look.

**Comment by Qinwroks** [comment]
:+1: 
@phouverneyuff great! 
and authoritites management is very important!

**Comment by phouverneyuff** [comment]
I think is important that we should use Resource instead of Roles in code. But this change would be made after this feature works.

Roles: ROLE_ADMIN, ROLE_USER
Resources: USER_CREATE, USER_EDIT, USER_DELETE, USER_VIEW, USER_LIST, USER_MANAGEMENT, USER_CHANGE_PASSWORD, ...

User have Role and Role have Resources.

Create, Edit and Remove User are resources, for example.

If the generator create 5 resources(CRUD) for each Entity and associate by default with ROLE_ADMIN, we can access the User Management or Role Management(in future) and associate the Resources to new Role.

The advantage of this is that we will not change the code if we need permit other user or a group of users for a specific feature.

@jdubois , @gzsombor @deepu105, @moifort what do you think?

**Comment by deepu105** [comment]
@phouverneyuff that would definitely be better than current role system coz in a real world application current role system is very difficult and manual to maintain

**Comment by moifort** [comment]
Sorry for late answer and the delay, I was trekking in the mountains this week and I have a lot of work too :disappointed:.

I manage to make it works with the default install (Java8&7, SQL), but not tested with MongoDB (I need to create a docker), and I try to develop for Cassandra.

The last commit contains some errors after generation, I will try to stabilize it today.

@Qinwroks when you say authorities management, its delete/create/modify an authority? If it is, I think is not a good idea, because if you delete/create/modify an authority you need to modify in the code too.

@phouverneyuff It's a very good idea :+1:!

**Comment by moifort** [comment]
I make a commit on my branch to fix the renaming and warning:
- Compile and Tested with Java7&8, SQL options
- Only compile with Java7&8 and MongoDB options

What remains to be done:
- Working with Cassandra
- Generate tests

**Comment by atomfrede** [comment]
I will again have a look and test it with mongodb options if you like.

**Comment by moifort** [comment]
@atomfrede yes! thank you! feel free to make any remarks!

**Comment by myspivey** [comment]
+1 this is an essential piece of any application that has a registration system

**Comment by Vaelyr** [comment]
I would suggest using User ID for queries instead of login name.

Also spotted a typo in javascript files for URL (/user-managment).

**Comment by myspivey** [comment]
UserManagementMapper also needs to exclude persistentTokens under the same rules as User

**Comment by moifort** [comment]
@myspivey @Vaelyr thanks for the feedback!

@myspivey fix!
@Vaelyr for the ID it was my first idea, but I take the lazy way, because `UserService` have a method to find an user only on the login, but if it's important I can change it. Fix the typo!

**Comment by Vaelyr** [comment]
Right, I made another method that accepts id and just used .findOne, I would just prefer following the restful convention using id's in paths, also prevents any encoding problems if login names contain non-English characters in which case the text is escaped and becomes a mess.

**Comment by phouverneyuff** [comment]
@moifort , I made some PRs on your fork. Take a look after please.

**Comment by ajaygodbole7** [comment]
@moifort, @phouverneyuff

This functionality is fantastic and kudos for building the prototype. Can't wait to see it in the main release.

However, I would caution against modelling resources. Since JHipster is deeply embedded with the Spring framework, the concept of Resource in Spring is already utilised.
http://docs.spring.io/spring/docs/current/spring-framework-reference/html/resources.html
Defining roles as part of security may lead to confusion and conflicts.

I would like to see JHipster to solve a complex role use-case using ACLs  as detailed here( pet clinic) : 
http://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#domain-acls

However, the ability to assign roles to users is a big leap forward.

**Comment by phouverneyuff** [comment]
@ajaygodbole7 my intention is use ACLs.

We need decide the name of the things now. I agree that resource and role isn't a good name.

Maybe: User > Profile > Permission

**Comment by ajaygodbole7** [comment]
@phouverneyuff

I think the concept of Roles is needed.

From my perspective,
Roles represent a set of permissions to take certain actions.
Users are granted a set of Roles.

So Users can take certain actions depending on Roles assigned to them.

Uers > Roles > Permissions

My Source: http://docs.aws.amazon.com/IAM/latest/UserGuide/roles-toplevel.html#roles-scenarios

Since JHipster's goal is to generate a modern web app based on angular and spring, we should leave the actual definition of what permissions mean to the designers of the app. These could be ACL based or JPA Entity based or Secure Objects or access to certain REST URIs etc.

In summary, what would be nice to have in JHipster would be:
1. Ability to assign Roles to Users.
2. Ability to create custom roles and assign them to Users.

**Comment by phouverneyuff** [comment]
@ajaygodbole7 I agree with Role. 
3. Ability to assign Permissions to Roles

**Comment by deepu105** [comment]
@phouverneyuff @ajaygodbole7 @moifort Yes complex real world role mapping seems very difficult to achieve with current architecture. let me give you a real world use case that im trying to tackle, The same is done today in JSP with JSTL and stuff successfully, but with angular and REST API seems very difficult
cc @jdubois @gzsombor 
* application has below structure from UI point of view
    * Modules (these are top level links in UI)
          * sub modules (A module can have one or many sub modules these are either tabs within a Page or second level menu in UI)
                * sections (These are logically grouped parts of a sub module)
* The application has users with following as attributes company, department, assignment and group
* Now the level of granular control is that users can have just view access to a section if they are in same department as the user who created the section data
* so I should be able to create a role and say which sections it has CRUD rights restricted to company, department, assignment or group, this restriction can be defined separately for each CRUD action

![image](https://cloud.githubusercontent.com/assets/1107223/8492645/6834331c-217f-11e5-81b2-eacadee0ef17.png)

**Comment by deepu105** [comment]
@phouverneyuff I haven't used spring security much before so do you think ACL can work for my above use case? if so do you have any suggestion?

**Comment by phouverneyuff** [comment]
I think that the page is so much attach(I don't if this the right word for this) with your business. If anything change, you'll code again. Try something more generally. Roles have Resources. Only this. Don't use table.

Try use a Picklist. If role have the resources, should be in right, if not in left.

Currently I work with Resources and Roles.
Security in view is for user experience, security in API is for your business and is mandatory.

I don't know if we'll use ACL yet. Maybe create two tables JHI_Resources and JHI_Roles_Resources can resolve the problem.

Try think modules as Resources.

**Comment by phouverneyuff** [comment]
e.g 
http://www.bootply.com/lFEZ9qOonX

**Comment by jdubois** [comment]
Hi guys,
Just don't use Spring Security ACLs. It just doesn't work:

- This has been available for years, and I've never seen anybody using it. Even while working at SpringSource.
- It's not even explained in the official Spring trainings
- I hate the way it's implemented, as you have your instances' permissions hard-coded in the database tables -> at least this used to be the case, I think it was refactored a few years ago, but I've never had a look at it, as once again nobody ever uses it

If you have something more complex than the usual "roles", then you normally have a business use case: you need to code your business rule, not use ACLs

**Comment by phouverneyuff** [comment]
Ok @jdubois . At night I'll do model the new tables. ( JHI_RESOURCES and JHI_ROLES_RESOURCES).

Can I Use MYSQL Workbench to Model and send you or do you prefer other IDE?

I'd like to change the name JHI_AUTHORITY for JHI_ROLES. Can I?

**Comment by moifort** [comment]
@phouverneyuff thanks a lot for your useful PR!
@Vaelyr you have a point! I will change it!
@atomfrede do you have the time to test with MongoDB?

**Comment by jdubois** [comment]
@phouverneyuff I don't understand why you want to add new tables, but I haven't followed this thread close enough, sorry. I will take the time to read it, but currently I'm trying to close older issues.
No don't change the JHI_AUTHORITY table name -> this would break stuff for too many people. Besides, the name comes from the default Spring Security schema, and even if it's not extremely good, we always try to be as close as possible to the libraries we use, so people aren't surprised.
@moifort you seem to be going quite fast -> tell me when you are ready, so I can test.
As I see you talk about ACL -> as a general rule, I'd rather have something simple, that I can review easily, and that people can start to use quickly. We can then improve it later. Better done than perfect!

**Comment by atomfrede** [comment]
@moifort Sorry didn't had time, will try it today or on sunday latest. But it looks like there are no problems in your code.

**Comment by phouverneyuff** [comment]
@jdubois , We want improve the permissions. But this is for future. Let's finish this PR first.

We want assign the permission instead of role(authorities) in code. It's more useful to code and more easy to maintain.

**Comment by Vaelyr** [comment]
Authority IS A permission. Role would be just a container that groups them. Only change in code would be that you loop through a role and assign the permissions that this role has to the Authentication object. It doesn't make anything better really.

**Comment by phouverneyuff** [comment]
If Authority is a permission, we shouldn't use ROLE_ADMIN and ROLE_USER. 

ROLE is like a Profile.

So we need create a way to group the permissions to facilitate the developer.

**Comment by moifort** [comment]
@jdubois not really fast :(, sorry for people who wait this functionality, I have lot of work right know. But thanks to @phouverneyuff  and @atomfrede for their help!

**Comment by deepu105** [comment]
@phouverneyuff i guess the sample you showed will be too difficult for a layman user to use, which is my case, so im looking from a usability and user experience perspective as well, I want the users to be able to configure complex roles easily. I believe that complexity should be on the code rather than in the UX thats the whole point of softwares isnt it make things easy for users. So I guess as @jdubois said i took a look at ACL and it really looks weird to me as well, they duplicate an entry for each entity, that doesnt seem good. I will end up having millions of entity instance in table and having individual rights for them doesnt seem good to me. So I guess I'll go ahead and do my own implementation for this. I guess this will not be good for Jhipster as this is more tied to my use case. so for JH your way of roles and resources looks good

**Comment by atomfrede** [comment]
@phouverneyuff How did you create a PR to @moifort repository? I would like to fix some things in the mongodb part.

**Comment by phouverneyuff** [comment]
@atomfrede yes
https://github.com/moifort/generator-jhipster/pulls
Please help us. lol

**Comment by atomfrede** [comment]
Ah thanks thats what I was looking for :)

**Comment by moifort** [comment]
I will need some help because I have no knowledge on MongoDB/Cassandra. I installed the docker MongoDB and Cassandra and try to run jhipster application but it's failed. Do I need to configure something on the docker? 

I use docker images: `tutum/mongodb` and `cassandra`

Error on mongoDB docker: `not authorized on jhipster to execute command { update: \"mongeez\"...`
Error on Cassandra docker: `com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed`

**Comment by atomfrede** [comment]
Why are you using mongodb inside docker? You can just run it locally. Docker makes it quite complicated. Same for cassandra (at least for Windows there is single instance binary).

**Comment by santhoshgit13** [comment]
Hi,

I am new to Angular JS and trying to check out the User Management screen. Can someone help me how to check out the code from github. When i look at the source itseems that i need to create the app similar to the default jhipster using command line but not sure how to do it from github.

Can anyone please help me with it.

Thanks

**Comment by atomfrede** [comment]
@moifort I've checked it out with cassandra (using local cassandra). Looks good so far. Following is not working right now: 
* Edit and display
* pagination seems to be enabled  by default on angular side (-> don't work)

**Comment by deepu105** [comment]
guys I have successfully implemented the role management for the usecase i explained above. I user spring security `AccessDecisionManager` and `AfterInvocationManager` along with `@Secured` annotation on resource methods. I used a custom annotation to provde restriction rules but thats not required for Jhipster. So @phouverneyuff I guess there is an easier way to do custom role creation and mapp it to users in Jhipster without using seperate persmissions, we can assign a grant(view/create/edit/delete) for each entity within a role profile and then use the spring security classes above to secure them based on users roles. This way we can have more user friendly UI to do role creation as well

**Comment by Vaelyr** [comment]
@deepu105 are you able to share your implementation?

**Comment by deepu105** [comment]
@Vaelyr not at this moment as it has some customer specific code inside but ill cleanup and post some snippets soon

**Comment by deepu105** [comment]
I have removed some specific business logic code also the `AfterInvocationManager ` may not be required. Its needed only if you want to filter returned objects with some logic
```
/**
 * Custom method security configuration
 * Created by Dee on 7/7/2015.
 */
@Configuration
@EnableGlobalMethodSecurity(prePostEnabled = true, securedEnabled = true, proxyTargetClass = true, mode = AdviceMode.PROXY)
@EnableAspectJAutoProxy
public class MethodSecurityConfiguration extends GlobalMethodSecurityConfiguration {

    public AccessDecisionManager accessDecisionManager() {
        CustomAccessDecisionManager accessDecisionManager = new CmsAccessDecisionManager();
        return accessDecisionManager;
    }

    public AfterInvocationManager afterInvocationManager() {
        CustomAfterInvocationManager afterInvocationManager = new CmsAfterInvocationManager();
        return afterInvocationManager;
    }

}
```

```
/**
 * Custom Access decision manager for method security
 * Created by Dee on 7/7/2015.
 */
public class CustomAccessDecisionManager implements AccessDecisionManager {

    private final Logger log = LoggerFactory.getLogger(CustomAccessDecisionManager .class);

    @Override
    public void decide(Authentication authentication, Object object, Collection<ConfigAttribute> configAttributes) throws AccessDeniedException, InsufficientAuthenticationException {
        ConfigAttribute configAttribute = SecurityUtils.getSingleConfigAttribute(configAttributes);

        MethodInvocation methodInvocation = (MethodInvocation) object;

        try {
            if(authentication != null) {
                User user = SecurityUtils.getCurrentUser(authentication);

                SecureRules secureRules = null;
                RequestMapping requestMapping = null;
                for (Annotation annotation : methodInvocation.getMethod().getDeclaredAnnotations()) {
                    if(annotation instanceof SecureRules){
                        secureRules = (SecureRules) annotation;
                    }else if(annotation instanceof RequestMapping){
                        requestMapping = (RequestMapping) annotation;
                    }
                }

                String method = SecurityUtils.getHttpMethod(secureRules, requestMapping);
                String sectionName = SecurityUtils.getReqPath(configAttribute);

                SecureRules.Rule[] rules = secureRules != null ? secureRules.rules() : new SecureRules.Rule[]{SecureRules.Rule.NONE};

                if (user != null && ! hasAccessToSection(cmsUser, method, sectionName ,methodInvocation, rules, configAttribute)) {
                    throw new AccessDeniedException("EX::User does not have enough privileges to access this method");
                }
            }

        } catch (ClassCastException e) {
            throw new IllegalStateException();
        }
    }

    @Override
    public boolean supports(ConfigAttribute attribute) {
        return false;
    }

    @Override
    public boolean supports(Class<?> clazz) {
        return clazz.equals(MethodInvocation.class);
    }
    /**
     * checks if the section has grants in any roles for the user
     * @param user
     * @param methodInvocation
     * @param rules
     *@param configAttribute  @return
     */
    private boolean hasAccessToSection(User user, String method, String sectionName, MethodInvocation methodInvocation, SecureRules.Rule[] rules, ConfigAttribute configAttribute) {

        Map<String, SectionDTO> validSections = user.getValidSections();
        if(validSections !=null && !validSections.isEmpty()) {
            SectionDTO validSection = validSections.get(sectionName);
            if (validSection != null && hasGrants(validSection, method) && isNotRestricted(user, validSection, rules, methodInvocation)) {
                log.debug("Role has access to section:{} ", validSection);
                return true;
            }
        }
        return false;
    }

    /**
     * checks for grant for HTTP methods
     * GET = 1 = VIEW,
     * POST = 2 = CREATE,
     * PUT = 3 = EDIT,
     * DELETE = 4 = DELETE
     *
     * @param section
     * @param method
     * @return
     */
    private boolean hasGrants(SectionDTO section, String method) {
        int grant = 0;
        switch (method){
            case "GET":
            case "VIEW":
                grant = 1; break;
            case "POST":
            case "CREATE":
                grant = 2; break;
            case "PUT":
            case "EDIT":
                grant = 3; break;
            case "DELETE":
                grant = 4; break;
            default:
                grant = 0;
        }

        return section.getGrant() >= grant;
    }

}
```

**Comment by Qinwroks** [comment]
@deepu105 really good job :+1:

**Comment by marcelinobadin** [comment]
Great! +1

**Comment by alchemistgo87** [comment]
Its a great feature. How soon will this feature be released?

**Comment by Falkurk** [comment]
Really great feature :+1: 

Is there a mean to fork your dev to use this feature now ?

**Comment by moifort** [comment]
Hello everybody!

First I am very sorry for the wait, I really dont have the time to work on it actually, I have a lot of work :(...

if you want my work this is my fork:

https://github.com/moifort/generator-jhipster/tree/user-managment
use the branch: user-management

**Comment by atomfrede** [comment]
@moifort If you're ok with it I can continue to work on it.

**Comment by tapas4java** [comment]
@moifort, Do you have a list of pending tasks in your mind?

**Comment by atomfrede** [comment]
Sql and MongoDB integration are working fine. Cassandra is missing.

**Comment by jdubois** [comment]
Cassandra isn't very important, very few people use it (even if I love it and I think that's a mistake!)

**Comment by tapas4java** [comment]
+1...We should not delay this feature for Cassandra support. Can we merge this PR now to master?

**Comment by atomfrede** [comment]
@tapas4java There is now PR at the moment, but I'll try to provide one asap

**Comment by andidev** [comment]
Great work! This will be really nice.

**Comment by moifort** [comment]
@atomfrede I can make the modification for disabling the functionality on Cassandra and make a pull request tonight or you prefer to do it?

**Comment by deepu105** [comment]
Great! Once this is done we can do role management

**Comment by atomfrede** [comment]
@moifort If you have time that would be faster I guess,

**Comment by conte7** [comment]
@moifort Can you help how to download the user-management .. please.... i newbie with this.... u.u

**Comment by moifort** [comment]
@conte7 
- The easy way: you can wait the new version of JHipster that contain the functionality and update your project or create a new one
- The difficult way: you make the change on your project reading the pull request  https://github.com/jhipster/generator-jhipster/pull/1930/files

**Comment by acheype** [comment]
Just a little fix to make it work when the user doesn't enable the translation support with Angular Translate. 

In the `user-management.js` file (`src/main/webapp/scripts/app/admin/user-management/user-management.js`), I had to delete manually the reference to the translatePartialLoader.

The file before :
```javascript
'use strict';

angular.module('malariaplantdbApp')
    .config(function ($stateProvider) {
        $stateProvider
            .state('user-management', {
                parent: 'admin',
                url: '/user-management',
                data: {
                    authorities: ['ROLE_ADMIN'],
                    pageTitle: 'Users'
                },
                views: {
                    'content@': {
                        templateUrl: 'scripts/app/admin/user-management/user-management.html',
                        controller: 'UserManagementController'
                    }
                },
                resolve: {
                    translatePartialLoader: ['$translate', '$translatePartialLoader', function ($translate, $translatePartialLoader) {
                        $translatePartialLoader.addPart('user-management');
                        return $translate.refresh();
                    }]
                }
            })
            .state('user-management-detail', {
                parent: 'admin',
                url: '/user-management/:login',
                data: {
                    authorities: ['ROLE_ADMIN'],
                    pageTitle: 'User'
                },
                views: {
                    'content@': {
                        templateUrl: 'scripts/app/admin/user-management/user-management-detail.html',
                        controller: 'UserManagementDetailController'
                    }
                },
                resolve: {
                    translatePartialLoader: ['$translate', '$translatePartialLoader', function ($translate, $translatePartialLoader) {
                        $translatePartialLoader.addPart('user-management');
                        return $translate.refresh();
                    }]
                }
            });
    });
```
And after :
```javascript
'use strict';

angular.module('malariaplantdbApp')
    .config(function ($stateProvider) {
        $stateProvider
            .state('user-management', {
                parent: 'admin',
                url: '/user-management',
                data: {
                    authorities: ['ROLE_ADMIN'],
                    pageTitle: 'Users'
                },
                views: {
                    'content@': {
                        templateUrl: 'scripts/app/admin/user-management/user-management.html',
                        controller: 'UserManagementController'
                    }
                },
                resolve: {

                }
            })
            .state('user-management-detail', {
                parent: 'admin',
                url: '/user-management/:login',
                data: {
                    authorities: ['ROLE_ADMIN'],
                    pageTitle: 'User'
                },
                views: {
                    'content@': {
                        templateUrl: 'scripts/app/admin/user-management/user-management-detail.html',
                        controller: 'UserManagementDetailController'
                    }
                },
                resolve: {

                }
            });
    });
```

**Comment by acheype** [comment]
And it seems to remain 4 references to the directive `translate-value-max` in the `user-management.html` file.

**Comment by flofly** [comment]
Hello, first of all, thanks for your work. Do you plan to add the ability to delete a user ?

**Comment by atomfrede** [comment]
Yes it's definitely planned.

**Comment by mallim** [comment]
May I know how to add user from User Management screen? Thanks for the good feature.

**Comment by Vaelyr** [comment]
Last time I checked it wasn't implemented fully yet, but would be rather easy to just create a new state for the detail view and reuse the registration logic that comes by default.

**Comment by sksandeep087** [comment]
Hello, I am Using Jhipster with Spring Boot 
1-problem is user-management list not showing, please give me a solution
![image](https://user-images.githubusercontent.com/34537939/39574776-f2978d5a-4ef5-11e8-8afc-fe2b77ecfa3d.png)

**Comment by jdubois** [comment]
@sksandeep087 please don't spam an old closed issue (3 years old!!). If you have a problem, please follow the guidelines.
